lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -4186,9 +4186,28 @@ bool SROA_Helper::LowerMemcpy(Value *V, DxilFieldAnnotation *annotation,
 
 /// MarkEmptyStructUsers - Add instruction related to Empty struct to DeadInsts.
 void SROA_Helper::MarkEmptyStructUsers(Value *V, SmallVector<Value *, 32> &DeadInsts) {
-  for (User *U : V->users()) {
-    MarkEmptyStructUsers(U, DeadInsts);
+  UndefValue *undef = UndefValue::get(V->getType());
+  for (auto itU = V->user_begin(), E = V->user_end(); itU != E;) {
+    Value *U = *(itU++);
+    // Kill memcpy, set operands to undef for call and ret, and recurse
+    if (MemCpyInst *MC = dyn_cast<MemCpyInst>(U)) {
+      DeadInsts.emplace_back(MC);
+    } else if (CallInst *CI = dyn_cast<CallInst>(U)) {
+      for (auto &operand : CI->operands()) {
+        if (operand == V)
```

Comment:
Minor, but I still think this code pattern is going to bite us one day when we encounter a `Call(V, V)` since I presume the Call will appear as a User twice, and each time you'll match either both or maybe the wrong operand. In this case it doesn't matter, but we should get into the habit of iterating over the llvm::Uses instead of the users if we're going to modify the Use relationship.